// @matterlabs/hardhat-zksync-node/dist/core/script-runner.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.runScriptWithHardhat = exports.runScript = void 0;
const execution_mode_1 = require("hardhat/internal/core/execution-mode");
const env_variables_1 = require("hardhat/internal/core/params/env-variables");
const path_1 = __importDefault(require("path"));
const utils_1 = require("../utils");
async function runScript(scriptPath, zksyncAnvilConfig, scriptArgs = [], extraNodeArgs = [], extraEnvVars = {}) {
    const { fork } = await Promise.resolve().then(() => __importStar(require('child_process')));
    const processExecArgv = withFixedInspectArg(process.execArgv);
    const nodeArgs = [
        ...processExecArgv,
        ...getTsNodeArgsIfNeeded(scriptPath, extraEnvVars.HARDHAT_TYPECHECK === 'true'),
        ...extraNodeArgs,
    ];
    const { commandArgs, server, port } = await (0, utils_1.startServer)(zksyncAnvilConfig.version, zksyncAnvilConfig.binaryPath, false, { quiet: true });
    await server.listen(commandArgs, false);
    await (0, utils_1.waitForNodeToBeReady)(port);
    const envVars = { ...process.env, ...extraEnvVars, ZKNodePort: port.toString() };
    return new Promise((resolve, reject) => {
        const childProcess = fork(scriptPath, scriptArgs, {
            stdio: 'inherit',
            execArgv: nodeArgs,
            env: envVars,
        });
        childProcess.once('close', async (status) => {
            await server.stop();
            resolve(status);
        });
        childProcess.once('error', async (error) => {
            await server.stop();
            reject(error);
        });
    });
}
exports.runScript = runScript;
async function runScriptWithHardhat(hardhatArguments, zksyncAnvilConfig, scriptPath, scriptArgs = [], extraNodeArgs = [], extraEnvVars = {}) {
    return runScript(scriptPath, zksyncAnvilConfig, scriptArgs, [...extraNodeArgs, '--require', path_1.default.join(__dirname, 'register')], {
        ...(0, env_variables_1.getEnvVariablesMap)(hardhatArguments),
        ...extraEnvVars,
    });
}
exports.runScriptWithHardhat = runScriptWithHardhat;
function withFixedInspectArg(argv) {
    const fixIfInspectArg = (arg) => {
        if (arg.toLowerCase().includes('--inspect-brk=')) {
            return '--inspect';
        }
        return arg;
    };
    return argv.map(fixIfInspectArg);
}
function getTsNodeArgsIfNeeded(scriptPath, shouldTypecheck) {
    if (process.execArgv.includes('ts-node/register')) {
        return [];
    }
    // if we are running the tests we only want to transpile, or these tests
    // take forever
    if ((0, execution_mode_1.isRunningHardhatCoreTests)()) {
        return ['--require', 'ts-node/register/transpile-only'];
    }
    // If the script we are going to run is .ts we need ts-node
    if (/\.tsx?$/i.test(scriptPath)) {
        return ['--require', `ts-node/register${shouldTypecheck ? '' : '/transpile-only'}`];
    }
    return [];
}
