// @matterlabs/hardhat-zksync-upgradable/dist/src/extension-generator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZkSyncGenerator = void 0;
const plugins_1 = require("hardhat/plugins");
const utils_1 = require("./utils");
class ZkSyncGenerator {
    constructor(_hre) {
        this._hre = _hre;
    }
    populateExtension() {
        this._hre.upgrades = (0, plugins_1.lazyObject)(() => this.makeFunctions());
        this._hre.zkUpgrades = (0, plugins_1.lazyObject)(() => this.makeFunctions());
    }
    makeFunctions() {
        const { makeDeployProxy } = require('./proxy-deployment/deploy-proxy');
        const { makeUpgradeProxy } = require('./proxy-upgrade/upgrade-proxy');
        const { makeValidateImplementation } = require('./validations/validate-implementation');
        const { makeDeployBeacon } = require('./proxy-deployment/deploy-beacon');
        const { makeDeployBeaconProxy } = require('./proxy-deployment/deploy-beacon-proxy');
        const { makeUpgradeBeacon } = require('./proxy-upgrade/upgrade-beacon');
        const { makeEstimateGasProxy } = require('./gas-estimation/estimate-gas-proxy');
        const { makeEstimateGasBeacon } = require('./gas-estimation/estimate-gas-beacon');
        const { makeEstimateGasBeaconProxy } = require('./gas-estimation/estimate-gas-beacon-proxy');
        const { makeChangeProxyAdmin, makeTransferProxyAdminOwnership } = require('./admin');
        return {
            deployProxy: (0, utils_1.wrapMakeFunction)(this._hre, makeDeployProxy(this._hre)),
            upgradeProxy: (0, utils_1.wrapMakeFunction)(this._hre, makeUpgradeProxy(this._hre)),
            validateImplementation: (0, utils_1.wrapMakeFunction)(this._hre, makeValidateImplementation(this._hre)),
            deployBeacon: (0, utils_1.wrapMakeFunction)(this._hre, makeDeployBeacon(this._hre)),
            deployBeaconProxy: (0, utils_1.wrapMakeFunction)(this._hre, makeDeployBeaconProxy(this._hre)),
            upgradeBeacon: (0, utils_1.wrapMakeFunction)(this._hre, makeUpgradeBeacon(this._hre)),
            admin: {
                changeProxyAdmin: (0, utils_1.wrapMakeFunction)(this._hre, makeChangeProxyAdmin(this._hre)),
                transferProxyAdminOwnership: (0, utils_1.wrapMakeFunction)(this._hre, makeTransferProxyAdminOwnership(this._hre)),
            },
            estimation: {
                estimateGasProxy: (0, utils_1.wrapMakeFunction)(this._hre, makeEstimateGasProxy(this._hre)),
                estimateGasBeacon: (0, utils_1.wrapMakeFunction)(this._hre, makeEstimateGasBeacon(this._hre)),
                estimateGasBeaconProxy: (0, utils_1.wrapMakeFunction)(this._hre, makeEstimateGasBeaconProxy(this._hre)),
            },
            // Properties from oz-upgrades
            forceImport: (0, utils_1.makeUndefinedFunction)(),
            silenceWarnings: (0, utils_1.makeUndefinedFunction)(),
            validateUpgrade: (0, utils_1.makeUndefinedFunction)(),
            deployImplementation: (0, utils_1.makeUndefinedFunction)(),
            prepareUpgrade: (0, utils_1.makeUndefinedFunction)(),
            beacon: {
                getImplementationAddress: (0, utils_1.makeUndefinedFunction)(),
            },
            erc1967: {
                getAdminAddress: (0, utils_1.makeUndefinedFunction)(),
                getImplementationAddress: (0, utils_1.makeUndefinedFunction)(),
                getBeaconAddress: (0, utils_1.makeUndefinedFunction)(),
            },
        };
    }
}
exports.ZkSyncGenerator = ZkSyncGenerator;
