// @matterlabs/hardhat-zksync-ethers/dist/hardhat-ethers/extension-generator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EthersGenerator = void 0;
const plugins_1 = require("hardhat/plugins");
class EthersGenerator {
    constructor(_hre) {
        this._hre = _hre;
    }
    populateExtension() {
        return (0, plugins_1.lazyObject)(() => {
            const hardhatEthersHelpers = require('@nomicfoundation/hardhat-ethers/internal/helpers');
            const { HardhatEthersProvider, } = require('@nomicfoundation/hardhat-ethers/internal/hardhat-ethers-provider');
            const { ethers } = require('ethers');
            const provider = new HardhatEthersProvider(this._hre.network.provider, this._hre.network.name);
            return {
                ...ethers,
                provider,
                getSigner: (address) => hardhatEthersHelpers.getSigner(this._hre, address),
                getSigners: () => hardhatEthersHelpers.getSigners(this._hre),
                getImpersonatedSigner: (address) => hardhatEthersHelpers.getImpersonatedSigner(this._hre, address),
                getContractFactory: hardhatEthersHelpers.getContractFactory.bind(null, this._hre),
                getContractFactoryFromArtifact: (artifact, signerOrOptions) => hardhatEthersHelpers.getContractFactoryFromArtifact(this._hre, artifact, signerOrOptions),
                getContractAt: (nameOrAbi, address, signer) => hardhatEthersHelpers.getContractAt(this._hre, nameOrAbi, address, signer),
                getContractAtFromArtifact: (artifact, address, signer) => hardhatEthersHelpers.getContractAtFromArtifact(this._hre, artifact, address, signer),
                deployContract: hardhatEthersHelpers.deployContract.bind(null, this._hre),
            };
        });
    }
}
exports.EthersGenerator = EthersGenerator;
