// @matterlabs/hardhat-zksync-ethers/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SUPPORTED_L1_TESTNETS = exports.ETH_DEFAULT_NETWORK_RPC_URL = exports.ZKVYPER_ARTIFACT_FORMAT_VERSION = exports.ZKSOLC_ARTIFACT_FORMAT_VERSION = exports.LOCAL_CHAINS_WITH_IMPERSONATION = exports.LOCAL_CHAIN_IDS = exports.LOCAL_CHAIN_IDS_ENUM = exports.PLUGIN_NAME = void 0;
exports.PLUGIN_NAME = '@matterlabs/hardhat-zksync-ethers';
// eslint-disable-next-line @typescript-eslint/naming-convention
var LOCAL_CHAIN_IDS_ENUM;
(function (LOCAL_CHAIN_IDS_ENUM) {
    LOCAL_CHAIN_IDS_ENUM["ERA_NODE"] = "0x104";
    LOCAL_CHAIN_IDS_ENUM["LOCAL_SETUP"] = "0x10e";
})(LOCAL_CHAIN_IDS_ENUM || (exports.LOCAL_CHAIN_IDS_ENUM = LOCAL_CHAIN_IDS_ENUM = {}));
exports.LOCAL_CHAIN_IDS = [LOCAL_CHAIN_IDS_ENUM.ERA_NODE, LOCAL_CHAIN_IDS_ENUM.LOCAL_SETUP];
exports.LOCAL_CHAINS_WITH_IMPERSONATION = [LOCAL_CHAIN_IDS_ENUM.ERA_NODE];
exports.ZKSOLC_ARTIFACT_FORMAT_VERSION = 'hh-zksolc-artifact-1';
exports.ZKVYPER_ARTIFACT_FORMAT_VERSION = 'hh-zkvyper-artifact-1';
exports.ETH_DEFAULT_NETWORK_RPC_URL = 'http://0.0.0.0:8545';
exports.SUPPORTED_L1_TESTNETS = ['mainnet', 'rinkeby', 'ropsten', 'kovan', 'goerli', 'sepolia'];
